### PR TITLE
feat: introduce new container authenticator

### PR DESCRIPTION
**Pull request by padamstx**
This commit introduces the new ContainerAuthenticator class,
along with associated tests and documentation.
The ContainerAuthenticator implements the authentication flow
to be used in IKS-managed compute resources where a secure compute resource token
has been injected into the local file system by the IKS compute resource provider.
The authenticator will read the cr token, then use it to obtain an IAM access token
by invoking the IAM "get token" operation with grant-type "cr-token".

### CODE REVIEW

File: Authentication.md [hunk 1]
```diff
@@ -1,9 +1,10 @@
 # Authentication
 The java-sdk-core project supports the following types of authentication:
 - Basic Authentication
-- Bearer Token
```
Comment (padamstx) [inline Authentication.md:4]: Some of the changes to this document are to track more closely the Go core version.

File: src/main/java/com/ibm/cloud/sdk/core/security/IamAuthenticator.java [hunk 2]
```diff
@@ -421,7 +421,8 @@ public IamToken requestToken() {
     // Form a POST request to retrieve the access token.
     RequestBuilder builder = RequestBuilder.post(RequestBuilder.resolveRequestUrl(this.getURL(), OPERATION_PATH));
 
-    // Now add the Content-Type and (optionally) the Authorization header to the token server request.
+    // Now add the Accept, Content-Type and (optionally) the Authorization header to the token server request.
+    builder.header(HttpHeaders.ACCEPT, HttpMediaType.APPLICATION_JSON);
```
Comment (padamstx) [inline src/main/java/com/ibm/cloud/sdk/core/security/IamAuthenticator.java:425]: I noticed that we weren't previously setting the Accept header on outbound IAM get-token requests, so added that here.  Apparently the IAM server doesn't care, but... :)

File: src/test/java/com/ibm/cloud/sdk/core/test/BaseServiceUnitTest.java [hunk 3]
```diff
@@ -80,4 +91,44 @@ protected static MockResponse jsonResponse(Object body) {
   protected static MockResponse errorResponse(int statusCode) {
     return new MockResponse().setResponseCode(statusCode);
   }
+
```
Comment (padamstx) [inline src/test/java/com/ibm/cloud/sdk/core/test/BaseServiceUnitTest.java:94]: Added a couple of utility methods to this test base class to be used by the authenticator tests.

File: src/test/java/com/ibm/cloud/sdk/core/test/security/ConfigBasedAuthenticatorFactoryTest.java [hunk 4]
```diff
@@ -45,37 +50,52 @@
     Map<String, String> env = new HashMap<>();
     env.put("SERVICE_1_URL", "https://service1/api");
     env.put("SERVICE_1_DISABLE_SSL", "true");
-    env.put("SERVICE2_URL", "https://service2/api");
```
Comment (padamstx) [inline src/test/java/com/ibm/cloud/sdk/core/test/security/ConfigBasedAuthenticatorFactoryTest.java:48]: Just consolidated some auth-related properties with their corresponding service-level properties, organized by service.

File: src/test/java/com/ibm/cloud/sdk/core/test/security/ConfigBasedAuthenticatorFactoryTest.java [hunk 5]
```diff
@@ -201,6 +278,12 @@ public void testEnvCredentialsService1() {
     Authenticator auth = ConfigBasedAuthenticatorFactory.getAuthenticator("service-1");
     assertNotNull(auth);
     assertEquals(Authenticator.AUTHTYPE_IAM, auth.authenticationType());
+    IamAuthenticator iamAuth = (IamAuthenticator) auth;
```
Comment (padamstx) [inline src/test/java/com/ibm/cloud/sdk/core/test/security/ConfigBasedAuthenticatorFactoryTest.java:281]: Noticed that we weren't doing much checking for the IAM and CP4D authenticators here so added some extra validation.

File: src/test/java/com/ibm/cloud/sdk/core/test/security/Cp4dAuthenticatorTest.java [hunk 6]
```diff
@@ -464,15 +463,6 @@ public void testApiErrorResponse() throws Throwable {
     }
   }
 
-  // Verify the Authorization header in the specified request builder.
```
Comment (padamstx) [inline src/test/java/com/ibm/cloud/sdk/core/test/security/Cp4dAuthenticatorTest.java:467]: Moved this to the base class.

**Review by dpopp07 — APPROVED** [review]
Looks good! 👍

**Pull request merged**